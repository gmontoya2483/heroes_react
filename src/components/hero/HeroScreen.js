import {useParams, Navigate, useNavigate} from "react-router-dom";
import {getHeroById} from "../../selectors/getHeroById";
import { useMemo } from "react";
import {heroImages} from "../../helpers/heroImages";
// import batman from '../../assets/dc-batman.jpg'  // Importación estática
// const heroImages = require.context('../../assets', true)



export const HeroScreen = () => {

    const navigate = useNavigate();
    const { heroeId } = useParams();

    const hero = useMemo(()=> getHeroById(heroeId),[ heroeId ]);

    if(!hero) {
        return <Navigate to={'/'} />
    }

    const {
        id,
        superhero,
        publisher,
        alter_ego,
        first_appearance,
        characters
    } = hero;


    const handleReturn = () => {
        navigate( -1 );
    }


    // const imagePath = `/assets/${ id }.jpg` // Public folder

    return (
        <div className={'row mt-5'}>
            <div className="col-4">
                <img
                    className={'img-thumbnail animate__animated animate__fadeInLeft'}
                    // src={ imagePath } // Public folder
                    // src ={ batman } // Importacion estática
                    src={ heroImages(`./${ id }.jpg`)}
                    alt={ superhero }
                />
            </div>

            <div className="col-8 animate__animated animate__fadeIn">
                <h3>{ superhero }</h3>
                <ul className={'list-group list-group-flush'}>
                    <li className={'list-group-item'}>
                        <strong>Alter ego:</strong> { alter_ego }
                    </li>

                    <li className={'list-group-item'}>
                        <strong>Publisher:</strong> { publisher }
                    </li>

                    <li className={'list-group-item'}>
                        <strong>First Appearance:</strong> { first_appearance }
                    </li>
                </ul>

                <h5 className={'mt-3'}>Characters</h5>
                <p>{ characters }</p>

                <button
                    className={'btn btn-outline-info'}
                    onClick={ handleReturn }
                >
                    Return
                </button>

            </div>


        </div>
    );
};
