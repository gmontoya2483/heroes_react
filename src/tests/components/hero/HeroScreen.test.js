import {mount} from "enzyme";
import {MemoryRouter, Route, Routes} from "react-router-dom";
import {HeroScreen} from "../../../components/hero/HeroScreen";
import {LoginScreen} from "../../../components/login/LoginScreen";



const mockNavigate = jest.fn();

jest.mock('react-router-dom', ()=> ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));


describe('Pruebas en <HeroScreen />', () => {


    test('no debe de mostrarse el HeroScreen sino hay un heroe en el URL', ()=> {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero']}>
                <Routes>
                    <Route path="/hero" element={<HeroScreen />}/>
                    <Route path="/" element={<h1>No Hero Page</h1>}/>
                </Routes>
            </MemoryRouter>

        );

        expect( wrapper.find('h1').text().trim()).toBe('No Hero Page')

    });


    test('debe de mostrarse el HeroScreen correctament', ()=> {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Routes>
                    <Route path="/hero/:heroeId" element={<HeroScreen />}/>
                    <Route path="/" element={<h1>No Hero Page</h1>}/>
                </Routes>
            </MemoryRouter>

        );

        expect( wrapper).toMatchSnapshot();
        expect( wrapper.find('h3').text().trim()).toBe('Spider Man')

    });


    test('debe de regrasar a la pantalla anterior', ()=> {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider']}>
                <Routes>
                    <Route path="/hero/:heroeId" element={<HeroScreen />}/>
                </Routes>
            </MemoryRouter>

        );

        wrapper.find('button').prop('onClick')();
        expect(mockNavigate).toHaveBeenCalledWith(-1);


    });



    test('No debe de mostrarse el HeroScreen si no lo encuentra', ()=> {

        const wrapper = mount(
            <MemoryRouter initialEntries={['/hero/marvel-spider123']}>
                <Routes>
                    <Route path="/hero/:heroeId" element={<HeroScreen />}/>
                    <Route path="/" element={<h1>No Hero Page</h1>}/>
                </Routes>
            </MemoryRouter>

        );

        expect( wrapper.find('h1').text().trim()).toBe('No Hero Page')

    });


});
