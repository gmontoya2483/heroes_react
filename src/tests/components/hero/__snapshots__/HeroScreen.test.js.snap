// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <HeroScreen /> debe de mostrarse el HeroScreen correctament 1`] = `
<div
  className="row mt-5"
>
  <div
    className="col-4"
  >
    <img
      alt="Spider Man"
      className="img-thumbnail animate__animated animate__fadeInLeft"
      src="/assets/marvel-spider.jpg"
    />
  </div>
  <div
    className="col-8 animate__animated animate__fadeIn"
  >
    <h3>
      Spider Man
    </h3>
    <ul
      className="list-group list-group-flush"
    >
      <li
        className="list-group-item"
      >
        <strong>
          Alter ego:
        </strong>
         
        Peter Parker
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          Publisher:
        </strong>
         
        Marvel Comics
      </li>
      <li
        className="list-group-item"
      >
        <strong>
          First Appearance:
        </strong>
         
        Amazing Fantasy #15
      </li>
    </ul>
    <h5
      className="mt-3"
    >
      Characters
    </h5>
    <p>
      Peter Parker
    </p>
    <button
      className="btn btn-outline-info"
      onClick={[Function]}
    >
      Return
    </button>
  </div>
</div>
`;
